lightning/src/ln/channelmanager.rs

```diff
@@ -4558,25 +4676,81 @@ impl<Signer: Sign, M: Deref, T: Deref, K: Deref, F: Deref, L: Deref> ChannelMana
 	/// will have the [`PaymentReceived::payment_preimage`] field filled in. That should then be
 	/// passed directly to [`claim_funds`].
 	///
 	/// See [`create_inbound_payment_for_hash`] for detailed documentation on behavior and requirements.
 	///
+	/// Note that a malicious eavesdropper can intuit whether an inbound payment was created by
+	/// `create_inbound_payment` or `create_inbound_payment_for_hash` based on runtime.
+	///
+	/// # Note
+	///
+	/// If you register an inbound payment with this method, then serialize the `ChannelManager`, then
+	/// deserialize it with a node running 0.0.103 and earlier, the payment will fail to be received.
+	///
 	/// [`claim_funds`]: Self::claim_funds
 	/// [`PaymentReceived`]: events::Event::PaymentReceived
 	/// [`PaymentReceived::payment_preimage`]: events::Event::PaymentReceived::payment_preimage
 	/// [`create_inbound_payment_for_hash`]: Self::create_inbound_payment_for_hash
+	// For details on the implementation of this method, see `verify_inbound_payment_data`.
 	pub fn create_inbound_payment(&self, min_value_msat: Option<u64>, invoice_expiry_delta_secs: u32) -> (PaymentHash, PaymentSecret) {
+		let min_amt_msat_bytes: [u8; 8] = match min_value_msat {
+			Some(amt) => amt.to_be_bytes(),
+			None => [0; 8],
+		};
+		// We assume that highest_seen_timestamp is pretty close to the current time - its updated when
+		// we receive a new block with the maximum time we've seen in a header. It should never be more
+		// than two hours in the future.  Thus, we add two hours here as a buffer to ensure we
+		// absolutely never fail a payment too early.
+		// Note that we assume that received blocks have reasonably up-to-date timestamps.
+		let expiry_bytes = (self.highest_seen_timestamp.load(Ordering::Acquire) as u64 + invoice_expiry_delta_secs as u64 + 7200).to_be_bytes();
+		let mut metadata_bytes: [u8; 16] = [0; 16];
+		{
+			let (min_amt_msat_slice, expiry_slice) = metadata_bytes.split_at_mut(8);
+			min_amt_msat_slice.copy_from_slice(&min_amt_msat_bytes);
+			expiry_slice.copy_from_slice(&expiry_bytes);
+		}
+
+		let rand_bytes = self.keys_manager.get_secure_random_bytes();
+		let iv_bytes = &rand_bytes[..16];
+
+		let mut hmac = HmacEngine::<Sha256>::new(&self.inbound_payment_key.ldk_pmt_hash_key);
+		hmac.input(iv_bytes);
+		hmac.input(&metadata_bytes);
+		let payment_preimage_bytes = Hmac::from_engine(hmac).into_inner();
+
+		let mut payment_secret_bytes: [u8; 32] = [0; 32];
+		{
+			let (iv_slice, encrypted_metadata_slice) = payment_secret_bytes.split_at_mut(16);
+			iv_slice.copy_from_slice(iv_bytes);
+
+			let chacha_block = ChaCha20::get_single_block(&self.inbound_payment_key.metadata_key, iv_bytes);
+			for i in 0..16 {
+				encrypted_metadata_slice[i] = chacha_block[i] ^ metadata_bytes[i];
+			}
+		}
+
+		let payment_hash = PaymentHash(Sha256::hash(&payment_preimage_bytes).into_inner());
+		(payment_hash, PaymentSecret(payment_secret_bytes))
+	}
```

Comment:
Between `create_inbound_payment` and `create_inbound_payment_for_hash` it seems like there's an opportunity to DRY up the code. Actually, along with `verify_inbound_payment_data`, too, there could be an abstraction used for construction / verification of each payment method, from which the payment hash and secret are pulled from. This would also be a natural place to move `ChaCha20::get_single_block` as per my earlier comment. Then all the logic is also encapsulated together.

Reply:
Good point, DRY'd `create_inbound_payment_*`.  Let me know what further abstraction you're thinking of.